caddyhttp/gzip/setup.go

```diff
@@ -119,3 +124,49 @@ func gzipParse(c *caddy.Controller) ([]Config, error) {
 
 	return configs, nil
 }
+
+// pool gzip.Writer according to compress level
+// so we can reuse allocations over time
+var (
+	writerPool             = map[int]*sync.Pool{}
+	defaultWriterPoolIndex int
+)
+
+func initWriterPool() {
+	var i int
+	for i = gzip.BestSpeed; i <= gzip.BestCompression; i++ {
+		writerPool[i] = &sync.Pool{
+			New: func() interface{} {
+				w, _ := gzip.NewWriterLevel(ioutil.Discard, i)
```

Comment:
There's a bug here.

`i` is closed over, so all pools (exception the default writer pool) will use the last value of `i` when called, which is 10 (gzip.BestCompression+1). Passing 10 to gzip.NewWriterLevel returns an error and a nil writer making the call to Reset in getWriter panic.

See https://play.golang.org/p/W9Chscrkme

Just declare a new variable `l := i` inside the loop and use it in Pool.New. This will fix the bug.

See https://play.golang.org/p/__42IlPQI8